### ISSUE TITLE
Screenshots are scaled for non-square pixel sources

### ISSUE THREAD

**Issue by etienned** [mpv-player/mpv#11857]
### Important Information

Screenshots are scaled for non-square pixel sources but should not, as is indicated in the documentation: _A screenshot will usually contain the **unscaled** video contents at the end of the video filter chain and subtitles._

Information:
- mpv version : 0.35.1
- Platform and Version : macos 12.6.6
- Source of the mpv binary : homebrew cask

### Reproduction steps

1. Open a video with non-square pixels (from command line or bundle app).
2. Take a screenshot (by pressing "s").
3. Compare the real pixel resolution of the video with the one of the image, they should be identical but they are not.

### Expected behavior
The screenshot should have the same resolution as the raw resolution of the video, not the same as the display resolution.

### Actual behavior
The screenshot have the same resolution as the displayed resolution of the video, not the same as the raw pixel resolution.

### Notes
For example, if I open a video file with non-square pixels with a resolution of 720x486, displayed at 720x540 (4:3), the screenshot have a resolution of 720x540 instead of 720x486. The problem was not present in previous versions but unfortunately I dont know when the change appears. Currently, as a workaround, I'm using `video-aspect-override=no` in the conf.

### Log file

[output.txt](https://github.com/mpv-player/mpv/files/11910375/output.txt)

**Comment by hooke007** [comment]
>  The problem was not present in previous versions

I tested 0.28.0 - 0.35.1
None of them would follow your description.

**Comment by Traneptora** [comment]
Try using `screenshot-sw=yes`.

**Comment by etienned** [comment]
I have tested `screenshot-sw=yes`, got the same result (source is a DV file, 720x480, screenshot is 720x540).

I have also tested with version [0.33.1](https://laboratory.stolendata.net/~djinn/mpv_osx/mpv-0.33.1.tar.gz), same result.

I have tested on Windows, version [0.34.0](https://sourceforge.net/projects/mpv-player-windows/files/64bit/mpv-x86_64-20220116-git-1ba0547.7z/download), also same result.

DV test file: https://sourceforge.net/projects/libdv/files/examples/examples/pond.dv/download?use_mirror=master&download=

**Comment by etienned** [comment]
It's desirable to keep exactly the frame data without altering it, for archival purposes for example or because the file will be process further and we don't want to loss too much quality by limiting the number of scaling generations. In our case, it's also desirable because we have a FFMpeg based pipeline to convert videos that also automatically crop small black or blurry borders of the video and do other things and this pipeline applied the same transformations to the screenshots, so they match exactly the video. For this to work, screenshots need to have the same resolution and pixel ratio.

Maybe 'Unscaled' means using the display resolution tagged in the source, but with older versions, it was really returning unscaled images in any ways, that's why we choose MPV in the first place. But, if you think that the current MPV have the desired behavior, it's fine and you can close this ticket.

**Comment by haasn** [comment]
Maybe we could add a new parameter to `screenshot` that would also disable anamorphic corrections?

It's a bit suspect to me to use screenshots in this way because it's not designed to be a lossless copy of the source one way or the other - between chroma upscaling, RGB conversion, bit depth conversions, lossy compression and other sources of distortion, anamorphic scaling feels more like the final nail in the coffin.

**Comment by Hrxn** [comment]
@etienned Have you looked into using the image output vo instead?

https://mpv.io/manual/master/#video-output-drivers-image

**Comment by etienned** [comment]
@haasn Adding a new parameter can be nice but is not mandatory, as I have a workaround with `video-aspect-override=no`. For the screenshot, I'm trying to reduce as much as possible all the losses and transformations (no lossy compression, control the color profile, etc.) and at the end is the result is relatively good, except for color accuracy, this one is tough.

@Hrxn Yes, but it doesn't really work with our workflow where someone has to select specific frames by watching the video. Thanks for the idea, though.

**Comment by Hrxn** [comment]
Yeah, but it can be used in conjunction with `--start` or `--end`, and maybe `--frames`

**Comment by etienned** [comment]
But it's still more complex than watching the video and hitting 's' to export the screenshot! And the people who do that here are not tech savvy, so no command line.

**Comment by Hrxn** [comment]
True, in that case.
But since you mentioned something about workflow..

**Comment by Dudemanguy** [comment]
Incidentally, I was just working on this for gpu-next. I don't believe this is a bug? The pixel resolution should be respected and match what is actually displayed. Is there something not sufficient about using `--video-aspect-override=no`?

**Comment by etienned** [comment]
@Dudemanguy At first I tough it was a bug because before (many versions ago) screenshots with non-square pixels were not scaled at all, but in recent versions it's no longer the case. So it's probably the older versions that had a bug (but this bug was satisfying my needs!). `--video-aspect-override=no` is sufficient for me, even if it's a little bit less satisfying for us because it force you to watch the video unscaled for non-square pixel videos.

**Comment by Dudemanguy** [comment]
You could set `--video-aspect-override=no` temporarily, take the screenshot, and change it back to `-1`. Unfortunately, taking a screenshot takes some time so it's not one simple keybind command input.conf. Maybe you could write a script that does this. I guess it's still not ideal though since the picture would fluctuate for that split second. Maybe it's worth having an option after all?

**Comment by pcordes** [comment]
There certainly should be some way to get a still image that contains the pixels the video file codes for.  If there isn't, that's at least a missing feature.

From the documentation, I'd certainly expect "unscaled" to mean truly unscaled, so image dimensions in pixels match the source video.

It might *also* sometimes be useful to get a jpeg or PNG that's been upscaled in on dimension or the other to have square pixels, but it seems to me that would be rarer.  If you want that, you can always un-fullscreen and take a shot of the image MPV is displaying in the window (without OSD).

As @etienned says, you want an image that matches the actual video pixels if you're going to use it to calculate args to `--vf=crop=...` or something (if `--vf=lavfi=cropdetect` doesn't do what you want).

Or if you're looking at individual pixels to see what distortion a video codec introduced, you want to see the actual pixels it coded for, not the results of some scaling algorithm MPV used which could smooth or sharpen, or exaggerate ringing artifacts.

**Comment by Anuskuss** [comment]
I ran into this issue today. I make heavy use of MKV's crop property, going to a bright scene in the video, screenshoting and then counting black pixels in GIMP. Only today I realized that MKV actually expects pixel values based on the original video dimensions (before squaring the pixels) which means I have to redo everything again.
I think screenshots should always be **unscaled** but I guess a `--screenshot-video-aspect-override=no` could also make sense. I'll run `--video-aspect-override=no` for the time being because I'm afraid that I might forget not to trust the default screenshots. It just sucks because actually watching the video is worse now.
